haddock/hscolour: fix highlighted source location

bennofs opened the pull request:
When generating haddocks with the `--for-hackage` switch, the generated
haddocks are placed in a different directory than the normal ones, which
includes the package id instead of just the package name. When we ran
hscolour, we didn't respect this, so the highlighted source would not
be placed in the correct directory and thus was missing from the tarball.
This patch fixes that.

Fixes #3451

bennofs commented:
Oops, seems like this still has a small bug. Please don't merge yet.

bennofs commented:
Fixed, should be ready to merge now.

23Skidoo commented:
@bennofs Apparently, this breaks the test suite. Can you look into that?

23Skidoo commented on Cabal/Distribution/Simple/BuildPaths.hs:
```diff
@@ -48,12 +49,18 @@ import System.FilePath ((</>), (<.>))
 srcPref distPref = distPref </> "src"
 
-hscolourPref :: FilePath -> PackageDescription -> FilePath
+data PathMode = ForDistribution | ForDevelopment
```
Needs a Haddock comment.

bennofs commented:
Alright, fixed that. I also renamed `PathMode` to `DistributionMode` since `PathMode` was quite generic (I was not able to come up with a better name than `DistributionMode` :|

23Skidoo commented:
If `ForDistribution` only gets set when `--for-hackage` is enabled, why not call it `ForHackage`? And then maybe we can change the type of `haddockForHackage` to `Flag DistributionMode` and perhaps also rename it to `haddockDistributionMode`.

bennofs commented:
Alright, that is a good idea. I'll do that.

bennofs commented:
Finally got around to finish this up. Should be ready now.

23Skidoo commented:
Merged, thanks!
